src/NumericalAlgorithms/LinearOperators/Divergence.hpp

```diff
@@ -7,98 +7,102 @@
+struct div;
+
+/// \cond
+template <typename Tag, typename Frame>
+struct div<Tag, Frame, Requires<tt::is_a_v<Tensor, db::item_type<Tag>>>>
+    : db::PrefixTag, db::SimpleTag {
+ private:
+  using first_index = tmpl::front<typename db::item_type<Tag>::index_list>;
+  static_assert(cpp17::is_same_v<typename first_index::Frame, Frame> and
+                    first_index::ul == UpLo::Up,
+                "First index of tensor cannot be contracted with derivative.");
+
+ public:
+  static std::string name() noexcept { return "div"; }
+  using tag = Tag;
+  using type = TensorMetafunctions::remove_first_index<db::item_type<Tag>>;
+};
+
+template <typename Tag, typename Frame>
+struct div<Tag, Frame, Requires<tt::is_a_v<::Variables, db::item_type<Tag>>>>
+    : db::PrefixTag, db::SimpleTag {
+  static std::string name() noexcept { return "div"; }
+  using tag = Tag;
+  using type = db::item_type<Tag>;
+};
+/// \endcond
 }  // namespace Tags
 
 /// \ingroup NumericalAlgorithmsGroup
 /// \brief Compute the (Euclidean) divergence of fluxes
-///
-/// \return a `Variables` with the same structure as `F` and each tag in
-/// `FluxTags` wrapped with a `Tags::div`.
 template <typename FluxTags, size_t Dim, typename DerivativeFrame>
 Variables<db::wrap_tags_in<Tags::div, FluxTags, DerivativeFrame>> divergence(
     const Variables<FluxTags>& F, const Index<Dim>& extents,
     const InverseJacobian<DataVector, Dim, Frame::Logical, DerivativeFrame>&
         inverse_jacobian) noexcept;
 
 namespace Tags {
-namespace Tags_detail {
-template <typename Tag, typename Frame>
-struct div_impl<Tag, Frame, Requires<tt::is_a_v<Tensor, db::item_type<Tag>>>>
-    : db::PrefixTag, db::SimpleTag {
-  static_assert(
-      cpp17::is_same_v<Frame,
-                       std::tuple_element_t<
-                           0, decltype(db::item_type<Tag>::index_frames())>>,
-      "the first index is not in the specified Frame");
-  using type = TensorMetafunctions::remove_first_index<db::item_type<Tag>>;
-  using tag = Tag;
-  static std::string name() noexcept { return "div"; }
-};
-
-template <typename... FluxTags, typename InverseJacobianTag>
-struct div_impl<tmpl::list<FluxTags...>, InverseJacobianTag,
-                Requires<tt::is_a_v<Tensor, db::item_type<InverseJacobianTag>>>>
-    : db::ComputeTag {
+/// \ingroup DataBoxTagsGroup
+/// \brief Compute the divergence of a Variables
+///
+/// Computes the divergence of the Tensors in the Variables
+/// represented by `Tag` in the frame mapped to by
+/// `InverseJacobianTag`.  The map must map from the logical frame.
+///
+/// This tag inherits from `db::add_prefix_tag<Tags::div, Tag, Frame>`
+/// with `Frame` the frame mapped to.
+template <typename Tag, typename InverseJacobianTag>
+struct ComputeDiv
+    : db::add_tag_prefix<
+          div, Tag,
+          std::tuple_element_t<
+              1, decltype(db::item_type<InverseJacobianTag>::index_frames())>>,
```

Comment:
@nilsdeppe out of curiosity (from Will and I) why does `index_frames()` return a tuple instead of a typelist?

Reply:
let's replace the `std::tuple_element...` with `tmpl::back<db::item_type<InverseJacobianTag>::index_list>::Frame` to make it consistent with the call to divergence below